src/interfaces/metadataemit.cpp

```diff
@@ -0,0 +1,3162 @@
+#include "metadataemit.hpp"
+#include "importhelpers.hpp"
+#include "signatures.hpp"
+#include "pal.hpp"
+#include <limits>
+#include <fstream>
+#include <stack>
+#include <algorithm>
+#include <utility>
+#include <cstring>
+
+#define RETURN_IF_FAILED(exp) \
+{ \
+    hr = (exp); \
+    if (FAILED(hr)) \
+    { \
+        return hr; \
+    } \
+}
+
+#define MD_MODULE_TOKEN TokenFromRid(1, mdtModule)
+#define MD_GLOBAL_PARENT_TOKEN TokenFromRid(1, mdtTypeDef)
+
+namespace
+{
+    void SplitTypeName(
+        char* typeName,
+        char const** nspace,
+        char const** name)
+    {
+        // Search for the last delimiter.
+        char* pos = std::strrchr(typeName, '.');
+        if (pos == nullptr)
+        {
+            // No namespace is indicated by an empty string.
+            *nspace = "";
+            *name = typeName;
+        }
+        else
+        {
+            *pos = '\0';
+            *nspace = typeName;
+            *name = pos + 1;
+        }
+    }
+}
+
+HRESULT MetadataEmit::SetModuleProps(
+        LPCWSTR     szName)
+{
+    // If the name is null, we have nothing to do.
+    // COMPAT-BREAK: CoreCLR would still record the token in the EncLog in this case.
+    if (szName == nullptr)
+        return S_OK;
+
+    pal::StringConvert<WCHAR, char> cvt(szName);
+    if (!cvt.Success())
+        return E_INVALIDARG;
+    
+    mdcursor_t c;
+    uint32_t count;
+    if (!md_create_cursor(MetaData(), mdtid_Module, &c, &count))
+    {
+        if (md_append_row(MetaData(), mdtid_Module, &c))
+        {
+            md_commit_row_add(c);
+        }
+        else
+        {
+            return E_FAIL;
+        }
+    }
+
+    // Search for a file name in the provided path
+    // and use that as the module name.
+    char* modulePath = cvt;
+    std::size_t len = std::strlen(modulePath);
+    const char* start = modulePath;
```

Comment:
Sorry. I'm sure this is going to suck. [East `const`](https://hackingcpp.com/cpp/design/east_vs_west_const.html)?